`azure_rm_adapplication_info` - Feature azure rm adapplication info diff

TheNotoriousRMM opened the pull request:
##### SUMMARY
azure_rm_adapplication_info add option "app_diff" #1559 

##### ISSUE TYPE
- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request

##### COMPONENT NAME
azure_rm_adapplication_info.py
test targets > azure_rm_adapplication > tasks: main

##### ADDITIONAL INFORMATION
```yaml
DOCUMENTATION = '''
options:
    app_diff:
        description:
            - The application Name or the app id.
        type: list
        elements: dict
```

TheNotoriousRMM committed:
modified:   plugins/modules/azure_rm_adapplication_info.py

TheNotoriousRMM committed:
modified:   plugins/modules/azure_rm_adapplication_info.py
app_diff hinzugefügt

TheNotoriousRMM committed:
modified:   plugins/modules/azure_rm_adapplication_info.py
Anpassungen

TheNotoriousRMM committed:
Anpassungen

TheNotoriousRMM committed:
Anpassungen

TheNotoriousRMM committed:
modified:   plugins/modules/azure_rm_adapplication_info.py
Finalizing

modified:   tests/integration/targets/azure_rm_adapplication/tasks/main.yml
Add tests for app_diff

TheNotoriousRMM committed:
modified:   plugins/modules/azure_rm_adapplication_info.py
modify doc for sentry test

TheNotoriousRMM committed:
Merge branch 'ansible-collections:dev' into feature_azure_rm_adapplication_info_diff

TheNotoriousRMM committed:
modified:   plugins/modules/azure_rm_adapplication_info.py
documentation modify examples

Fred-sun commented on tests/integration/targets/azure_rm_adapplication/tasks/main.yml:
```diff
@@ -46,45 +43,58 @@
+
+- name: Assert the difference
+  ansible.builtin.assert:
+    that: diff_output.app_diff[0].app_display_name == "{{ display_name }}"
```
Wrong indentation: expected 6 but found 8

TheNotoriousRMM replied:
I don't understand this issue..
The indentaton is the same as in line 84, 85 and 86?

Fred-sun commented on tests/integration/targets/azure_rm_adapplication/tasks/main.yml:
```diff
@@ -94,4 +104,4 @@
   ansible.builtin.assert:
     that:
-      - output.applications | length == 0
+      - output.applications | length == 0
```
No new line character at the end of file

TheNotoriousRMM replied:
New commit: Fixed in 1 line.

Fred-sun commented on plugins/modules/azure_rm_adapplication_info.py:
```diff
@@ -149,7 +229,8 @@ def __init__(self):
             identifier_uri=dict(type='str'),
-            app_display_name=dict(type='str')
+            app_display_name=dict(type='str'),
+            app_diff=dict(type='list', elements='dict')
```
If defined this way, it will be easier to understand!
```suggestion
            app_diff=dict(type='list',
                                   elements='dict',
                                   options=dict(
                                                         app_id=dict(type='str'),
                                                         app_display_name=dict(type='str')
                                   
                                   )
            )
```

TheNotoriousRMM replied:
If I have the configuration of my adapps in a YAML or JSON file, I don't want to edit it additionally. I want to provide it directly to the plugin, not just the "app_display_name" or the "app_id". However, if I provide this as an option, I have to filter the YAML or JSON for these two options, which means another Ansible task in the playbook.

Fred-sun replied:
But your code only handles these two fields and no other fields?

TheNotoriousRMM replied:
My code need one of these fields, the other fields will be ignored. In the future with this way, it's possible to check the diffrence of any field and to update them.
So I can write the full config for the AD App by code, check it with the app_diff option and create, update or delete the applications with the existing module "azure_rm_adapplication.py".

therapac replied:
Yes, that is correct. The code only handles these two fields. However, I can provide a YAML or JSON that contains the entire content of multiple ADAPPs in a list without it being blocked. This way, I can directly compare my stored configuration of the ADAPPs without having to parse it again, create a list, and pass the values into a variable

Fred-sun commented on plugins/modules/azure_rm_adapplication_info.py:
```diff
@@ -183,12 +263,27 @@ def exec_module(self, **kwargs):
+                    if app.get("app_id") == diff.get("app_id") or app.get("app_display_name") == diff.get("app_display_name"):
+                        found = True
+                        break
+                if not found:
```
Is this a definition error? Add to 'app_diff' if it should be True! Thank you!
```suggestion
                if found:
```

TheNotoriousRMM replied:
No, that is correct. It compares the provided list with the production Azure data, and if something is not present in the production environment, the status "absent" is added. This way, you can immediately reuse the return to delete the adapps that do not belong in the production environment using "plugins/modules/azure_rm_adapplication.py". 
This allows you to first add adapps with "plugins/modules/azure_rm_adapplication.py" and then use "plugins/modules/azure_rm_adapplication_info.py" (option diff) to identify the adapps that should not be in the live environment and subsequently delete them with "plugins/modules/azure_rm_adapplication.py".

Fred-sun replied:
Understood, please fix the following error, and make a detailed definition of the parameter definition. Thank you!

therapac replied:
I have adjusted the documentation and re-added the NOT.

Fred-sun reviewed (commented):
@TheNotorioursRMM Please take a look, Small change request! Thanks!

Fred-sun commented:
kindly ping!

TheNotoriousRMM committed:
modified:   tests/integration/targets/azure_rm_adapplication/tasks/main.yml
Fix: No new line character at the end of file

Fred-sun commented:
@TheNotoriousRMM Please help solve the conflicts  I will review and push forward the merger as soon as possible. Thank you!

TheNotoriousRMM committed:
Merge branch 'dev' into feature_azure_rm_adapplication_info_diff

Fred-sun commented on tests/integration/targets/azure_rm_adapplication/tasks/main.yml:
```diff
@@ -103,5 +120,4 @@
   ansible.builtin.assert:
-    that:
-      - output.applications | length == 0
+    that: output.applications | length == 0
```
No new line character at the end of file

Fred-sun replied:
This error still exist, No new line character at the end of file
```suggestion
    that:
      output.applications | length == 0
```

Fred-sun commented on plugins/modules/azure_rm_adapplication_info.py:
```diff
@@ -63,6 +68,12 @@
+- name: get ad app diff ---- by display name
+  azure_rm_adapplication_info:
+    app_diff:
+        - app_display_name: "{{ display_name }}"
```
Wrong indentation: expected 6 but found 8

therapac replied:
Fix wrong indentation

Fred-sun commented on plugins/modules/azure_rm_adapplication_info.py:
```diff
@@ -63,6 +68,12 @@
+  azure_rm_adapplication_info:
+    app_diff:
+        - app_display_name: "{{ display_name }}"
+        - app_id: "{{ app_id }}"
```
Wrong indentation: expected 6 but found 8

Fred-sun reviewed (commented):
small change request!

therapac committed:
modified:   plugins/modules/azure_rm_adapplication_info.py
Add new documentation for the option app_diff

therapac committed:
modified:   plugins/modules/azure_rm_adapplication_info.py
Fix wrong indentation

therapac committed:
Merge branch 'ansible-collections:dev' into feature_azure_rm_adapplication_info_diff

Fred-sun commented on tests/integration/targets/azure_rm_adapplication/tasks/main.yml:
```diff
@@ -103,5 +120,4 @@
   ansible.builtin.assert:
-    that:
-      - output.applications | length == 0
+    that: output.applications | length == 0
```
The error( No new line character at the end of file)  still exsit, Please '
```suggestion
    that:
      output.applications | length == 0
```

Fred-sun reviewed (commented):
Please!

TheNotoriousRMM committed:
Update tests/integration/targets/azure_rm_adapplication/tasks/main.yml

I hope this will resolve the issue.

Co-authored-by: Fred-sun <37327967+Fred-sun@users.noreply.github.com>

TheNotoriousRMM reviewed (commented):
Change accepted

Fred-sun commented on tests/integration/targets/azure_rm_adapplication/tasks/main.yml:
```diff
@@ -104,4 +121,4 @@
   ansible.builtin.assert:
     that:
-      - output.applications | length == 0
+      output.applications | length == 0
```
Same error!
```suggestion
  ansible.builtin.assert:
    that: output.applications | length == 0
```

TheNotoriousRMM replied:
Resolved in a new push

TheNotoriousRMM committed:
Merge branch 'ansible-collections:dev' into feature_azure_rm_adapplication_info_diff

TheNotoriousRMM committed:
modified:   tests/integration/targets/azure_rm_adapplication/tasks/main.yml
Change No new line character at the end of file

TheNotoriousRMM reviewed (commented):
I have fix line 124

Fred-sun commented:
@TheNotoriousRMM sanity tests still don't work, depending on the editor you're using. The error is as follows:
```e.g.
yaml[new-line-at-end-of-file]: No new line character at the end of file
tests/integration/targets/azure_rm_adapplication/tasks/main.yml:123

yaml[new-line-at-end-of-file] basic   formatting, yaml

Fred-sun commented:
@TheNotoriousRMM Since you did not make any changes in the last line, can you try to undo the changes in the last line? Thank you!

TheNotoriousRMM committed:
modified:   tests/integration/targets/azure_rm_adapplication/tasks/main.yml
Undo changes at the last line. Sanity test was OK.

TheNotoriousRMM reviewed (commented):
I have undo the changes at the last line. Sanity check was OK.

Fred-sun commented:
@TheNotoriousRMM However, judging from the PR content you submitted, you added new lines at the end of the play, as shown below:
![image](https://github.com/ansible-collections/azure/assets/37327967/d81afa96-0838-46f4-8449-d76e1958afec)

TheNotoriousRMM commented:
> @TheNotoriousRMM However, judging from the PR content you submitted, you added new lines at the end of the play, as shown below: ![image](https://private-user-images.githubusercontent.com/37327967/343815912-d81afa96-0838-46f4-8449-d76e1958afec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk1NjEyMDIsIm5iZiI6MTcxOTU2MDkwMiwicGF0aCI6Ii8zNzMyNzk2Ny8zNDM4MTU5MTItZDgxYWZhOTYtMDgzOC00NmY0LTg0NDktZDc2ZTE5NThhZmVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI4VDA3NDgyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRlNDljM2YyYjA5NWFjYmQ1YjY0YjhlYTBhM2VmOGQ3NGRjODgyM2VjOWE3NTYyZDcxNzYzODFjNzE1MTc0ODkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.37fikVHsWn4qOVVZceEasHd3k9CJL-tORMg-2wXQ0AY)

I'm confused.. Can you give me a hint how I can delete this line?
In my GitHub Repo I have no new line at the end of the file:
![image](https://github.com/ansible-collections/azure/assets/164023381/22d0db3e-b1a3-4008-a8be-02bebbcfe7ee)

Fred-sun commented:
@TheNotoriousRMM You can avoid this problem by moving the mouse over the last line, pressing the 'ESC' key to cancel the new line at the tail and then save, I did this in PR #1616 and the tail line disappeared, sanity pass! Thank you!

TheNotoriousRMM committed:
modified:   tests/integration/targets/azure_rm_adapplication/tasks/main.yml
Add new line at the end of the file. Opened in vim and save, the new line automaticly added.

TheNotoriousRMM commented:
@Fred-sun I have finaly added a new line at the end of the file. Sorry, I thought the last line had to be deleted and not added. I didn't understand you correctly ;)
Hope the file looks good now.

Fred-sun commented:
@TheNotoriousRMM Thanks for you update! Everything is OK! I will push for merged as soon as possible!
